spec/treasury_and_analytics.md

```diff
@@ -0,0 +1,135 @@
+# Treasury and Analytics
+
+This specification provides details of BitDAO treasury and analytics pages, as well as associated
+required services.
+
+We first describe how services are constructed, then we describe how the data is presented to users.
+
+The treasury and analytics approach taken here is based on daily values. This specification is not
+intended to serve shorter periods or live updates (the backend services will need to be re-evaluated
+for performance and flexibility, and likely requires rebuilding).
+
+## Features
+
+- Idempotent pipeline
+- Static page generation
+- Serverless using object storage, i.e., no key-value stores or relational database administration
+- Decentralized analytics by enabling self-contained pages
+
+## Analytics
+
+Data retrieval:
+
+- A serverless code executable retrieves token quantities and spot prices once a day.
+- Retrieved values are saved to S3 (as CSV or JSON) using the following filename convention:
+    - {token}-{yyyy}-{mm}-{dd}.json- If spot prices are pulled from multiple sources, use the
+      following filename convention:
+- Optional: Retrieve spot prices from multiple sources and store in the same file
+
+#### Token quantity specification:
+
+| Field | Type | Description |
+| ---- | ---- | ---- |
+| yyyy | number | Year of retrieved value: MUST equal 2021 | 
+| mm | number | Month of retrieved value: MUST be between 1 and 12 | 
+| dd | number | Day of retrieved value: MUST be between 1 and 31 (validation tbd) |
+| token | String | Short token identifier, e.g., BIT, BTC, ETH, USDC, USDT | 
+| quantity | BigInt | Number of tokens: MUST be greater than or equal to 0: MUST support BigInt |
+| source | String | The url where this row's value was obtained, e.g., https://coinmarketcap.com/currencies/bitdao/ | 
+
+#### Spot price specification:
+
+| Field | Type | Description |
+| ---- | ---- | ---- |
+| yyyy | number | Year of retrieved value: MUST equal 2021 |
+| mm | number | Month of retrieved value: MUST be between 1 and 12 |
+| dd | number | Day of retrieved value: MUST be between 1 and 31 (validation tbd) |
+| token | String | Short token identifier, e.g., BIT, BTC, ETH, USDC, USDT |
+| price | Decimal | Price of token: must accommodate exactly 2 decimal places, greater than 0, less than 5,000 |
+| source | String | The url where this row's value was obtained, e.g., https://coinmarketcap.com/currencies/bitdao/ | 
+
+## Building daily static page
+
+Procedure to build static page:
+
+1. Retrieve token spot price and store on S3
+   1. Copy content of existing data file and store with a new name
+2. Retrieve token amounts from BitDAO treasury
+    1. Copy content of existing data file and store with a new name
+3. Generate the static html page with links pointing to the latest data files
+4. Include analytics javascript library to allow online aggregation 
+
+## Treasury, Holdings
+
+Display current ERC-20 token balance of the following coins:
```

Comment:
suggestion: DAO provides a callable function listing the assets (token abbreviations returned as a list) under management, (held by the DAO, not any sub-treasuries).

Reply:
Sounds like a reasonable approach, though it'll require admin/multi-sig involvement. It might be easy enough to maintain off-chain since the pledged contribution is well-defined and is handled off-chain, i.e., we know the sources we draw from handle the tokens we want to show, which may become an issue if a different token shows up all of a sudden.

Reply:
sort of related of what I don't like on this api here https://www.bitdao.io/api/service/chart?day=100 there is not an strict pair of proprieties to do a loop on the code and display it like: 
[{ token: eth, value: 1, valueUsd:3400 },{}] 
then the front end will just loop what ever is on the array just waiting to have for example data.token data.value on the template.

Reply:
This specification is only for the current BitDAO treasury and will be adapted to cover sub-treasuries/sub-DAOs when we have some additional clarity and structure about sub-treasuries (likely, starting with the bond contract). 

For now, the token list of each DAO should be explicitly specified rather than querying an online feed and reporting all the tokens they return. One of the reasons why is because currently there are some spurious tokens that have been sent to the treasury which need to be managed out of the treasury - they aren't part of the bybit pledge, and no agreements are in place between that token and the BitDAO.